### DISCUSSION TITLE
Adopt JSDoc to document TypeScript code

### DISCUSSION THREAD

**Discussion by IuryPiva** [bokeh/bokeh#11309] [Documentation]
It would be useful to add [JSDoc](https://jsdoc.app) to Bokeh's TypeScript code, having in mind that some code is unique to the TypeScript part of Bokeh and it's getting quite big and complex at the moment. Additionally, we can use tools like [sphinx-js](https://pypi.org/project/sphinx-js/) to increment our documentation.

cc: @tcmetzger @bryevdv @mattpap

**Comment by mattpap** [comment]
Given bokehjs is written in TypeScript, any tool of choice should be TypeScript specific, e.g. https://typedoc.org/. Last time I tried it, it had issues with our codebase, but hopefully it improved by now.

**Comment by bryevdv** [comment]
There are a couple of things to consider. Primarily, who is the audience for "API" docs?  Setting aside `bokehjs-api` which definitely needs TypeDoc-style API docs, most *user* usage of BokehJS in `CustomJS` only involves setting property values (e.g. `source.data = new_data`) and almost never involves *calling functions*. In fact off the top of my head, pretty much the only function I can think of that users might call regularly in `CustomJS` is variants of `change.emit()` In that sense, the existing Python reference guide has all of the relevant information. Perhaps this fact that the properties are mirrored on both sides could be made more clear. 

Accordingly, I think doctrings for methods/functions in the library would be mostly for developers of Bokeh itself. I think that's fine but we should keep that in mind explicitly, both when prioritizing and deciding where and how to publish the docs, and what context to frame them in. 

So then: is there a place for primarily *user-facing* BokehJS docs? I think there is one existing, and one potential:

The first is `bokehjs-api` as mentioned earlier, this is explicitly a function API intended for users to use direclty. No question it should get docstrings and published API docs.

What also might be relevant in the future is if we decide to create APIs specifically to be useful in `CustomJS` (and/or extensions) for various common tasks in those contexts. If we start making some APIs for this purpose we could potentially insulate `CustomJS` and extension code from "internal" changes to BokehJS to a greater degree than we are currently able.
